[#148] Scripts and Queries for Database Population

TiraFire opened the pull request:
Within this PR, there are the builders which scrape the website and turn the information into JSON datasets, another builder to use the dataset text files with the array points to create JSON datasets with datapoints, and another builder to use the citations from the website and create JSON authors and publications. 

Next are the scripts to use these JSON files to create the queries for each of the tables. The numbers on the file correspond to which script to run first (there is a small issue concerning the subcat, and cat that requires order)

Finally, there are the queries in MySQL format that also have an order to follow to prevent FK problems.

JeremiahTee committed:
[#148] Initial builder scripts

To be updated by Jessica

JeremiahTee committed:
[#148] authors, material, publications query .txt files

These were used to insert into databoom_db_test initially

JeremiahTee committed:
[#148] JSON files produced by builders and scripts

PAA refers to publication-authors-author table.

JeremiahTee committed:
[#148] JSON files produces by builders and scripts

PAA refers to publication-authors-author table.

JeremiahTee committed:
[#148] Authors, composition scripts to build query .txt files

composition-script produces 
- materials-query.txt
- dataset-materials-material.txt
- composition-query.txt

JeremiahTee committed:
[#148] Authors, composition scripts to build query .txt files

composition-script produces 
- materials-query.txt
- dataset-materials-material.txt
- composition-query.txt

JeremiahTee committed:
[#148] Composition query .txt file produced by its corresponding script

composition-query.txt file is produced by composition-script.py

This contains all the queries to populate composition table.

JeremiahTee committed:
[#148] Composition query .txt file produced by its corresponding script

composition-query.txt file is produced by composition-script.py

This contains all the queries to populate composition table.

JeremiahTee committed:
Merge remote-tracking branch 'origin/Database/Populate-DB' into Database/Populate-DB

TiraFire committed:
[#148] Additional python scripts for JSON building and Queries

Added the missing python scripts for the citations breakdown into the PAA, authors and publications JSON files, as well as the missing script for the datasets with datapoints json

TiraFire committed:
Merge remote-tracking branch 'origin/Database/Populate-DB' into Database/Populate-DB

JeremiahTee committed:
[#148] file renames and add comments to each script

JeremiahTee committed:
[#148] change to correct paths and run all scripts to generate queries

JeremiahTee committed:
[#148] rename folder for txt files with datapoints 

category-subcat-query contains the default queries so those 2 queries are removed from default-queries,txt

JeremiahTee committed:
[#148] name script and queries with order, modify builders

JeremiahTee committed:
[#148] add script to generate units, repr, datapoint queries

JeremiahTee committed:
[#148] run all scripts in order

by running the category script, it generates the categoryId and subcategoryId on datasets-with-datapoints.json

JeremiahTee committed:
remove quotes from galloping

this is so it can execute

TiraFire committed:
[#148] Possible solution to Array issue problem with datapoint queries

JeremiahTee committed:
[#148] comments on script, final touches

- Removed double quotes from "Galloping" in citations.txt
- We tested foreign key checks statements from every queries and script, because we tested the SCRIPTS and QUERIES in order. They work beautifully :)

TiraFire committed:
[#148] Slight change to cat/subcat to add the IDs manually

TiraFire committed:
[#148] Added final comments to the Author_Publications builder

TiraFire committed:
[#148] Fixed some file locations, test to correct db, corrected...

a citation issue for id:39. Should be good to go :)

JeremiahTee committed:
correct naming of 7-units-repr query text file

JeremiahTee committed:
Merge remote-tracking branch 'origin/Database/Populate-DB' into Database/Populate-DB

JeremiahTee commented:
As part of the #148 user story, I have written a detailed report of all the work on Confluence. I have also written an article describing the scrapers that build datasets.json and datasets-with-datapoints.json.

[**[Report] Scrapers to build datasets.json and datasets-with-datapoints.json**](https://databoom.atlassian.net/wiki/spaces/DB/pages/122814761/Report+Scrapers+to+build+datasets.json+and+datasets-with-datapoints.json)

This article describes all the work from this PR in detail. I highly recommend reading it.
[**[Report] How To Populate Database using scraped JSONs**](https://databoom.atlassian.net/wiki/spaces/DB/pages/170557454/Report+How+To+Populate+Database+using+scraped+JSONs)


(screenshots to follow)

JeremiahTee commented:
Screenshots from [**[Report] Scrapers to build datasets.json and datasets-with-datapoints.json**](https://databoom.atlassian.net/wiki/spaces/DB/pages/122814761/Report+Scrapers+to+build+datasets.json+and+datasets-with-datapoints.json) 
Figure 1: View of the table contents 
![image](https://user-images.githubusercontent.com/6520150/103845702-9155b200-506a-11eb-8094-cbfa394253c7.png)
Figure 2: View of the **datasets-builder.py purpose** section
![image](https://user-images.githubusercontent.com/6520150/103845751-af231700-506a-11eb-8341-baa33f6c215a.png)

Figure 3: View of the **datasets-builder.py breakdown & output** section
![image](https://user-images.githubusercontent.com/6520150/103845785-c5c96e00-506a-11eb-9076-4bb8a1a1a199.png)

JeremiahTee commented:
Screenshots from 
[**[Report] How To Populate Database using scraped JSONs**](https://databoom.atlassian.net/wiki/spaces/DB/pages/170557454/Report+How+To+Populate+Database+using+scraped+JSONs)

Figure 4: Table of contents of article
![image](https://user-images.githubusercontent.com/6520150/103846029-4b4d1e00-506b-11eb-99ab-a668b1c67c5b.png)

Figure 5: View of **Steps of how we populated the database** section
![image](https://user-images.githubusercontent.com/6520150/103845907-0923dc80-506b-11eb-8a04-d50c9ec7aba4.png)

Figure 6: View of Breakdown of Database/Scripts folder
![image](https://user-images.githubusercontent.com/6520150/103845975-307aa980-506b-11eb-8b03-bda27c4d5041.png)

JeremiahTee commented:
Finally, I'd like to include a screenshot from the last section in the [**[Report] How To Populate Database using scraped JSONs**](https://databoom.atlassian.net/wiki/spaces/DB/pages/170557454/Report+How+To+Populate+Database+using+scraped+JSONs) article. This table indicates all the scripts, queries text files and JSON used.

Figure 7:
![image](https://user-images.githubusercontent.com/6520150/103846181-8fd8b980-506b-11eb-84cb-e897cbaa06e6.png)

JeremiahTee commented:
As a side note, we will create an email account for the teacher to visit the Confluence and verify the pages at his own liking. @TiraFire @GD-Dheer

silent-po2 reviewed (approved):
This was a large and time consuming undertaking, longer than what was initially expected. However, this is all excellent work and a complete replication/replacement of the old website's DB. Looking over the values entered in our DB, I could only find a single duplicate entry in the authors table; easily fixable. There were no duplicates elsewhere, which is perfect. Well done.

GD-Dheer commented on Database/Scripts/Builders/txt with datapoints/at108a.txt:
```diff
@@ -0,0 +1,5 @@
+#Equivalence Ratio, Critical Charge (g tetryl), Critical Energy (J)
+0.78157, 9.94, 44932.0
```
How are we getting this data? What is difference between this data and the data on the old caltech site?

TiraFire replied:
These texts files were scraped from the caltech website by Jamal, the only issue is that there are some of these files that needed to be changed manually to work with our scripts (to prevent a considerable amount of if statements for the small amount of outliers). These changes are small such as headers, missing punctuation, improper units, etc

JeremiahTee replied:
As discussed, @TiraFire will note these changes in the Confluence article
[**[Report] How To Populate Database using scraped JSONs**](https://databoom.atlassian.net/wiki/spaces/DB/pages/170557454/Report+How+To+Populate+Database+using+scraped+JSONs)

GD-Dheer commented on Database/Scripts/Builders/datapoints-builder.py:
```diff
@@ -0,0 +1,97 @@
+import json
+
+with open('./Scraped JSONs/datasets.json') as f:
```
Your scripts are creating these JSON files and then you are feeding them into the next script. In that case, is there a reason we need to commit /Scraped JSONS? We can let the user recreate these if required

TiraFire replied:
We can remove them in that case but for the purpose of helping others, keeping them would mean they can access the JSON files without needing to run the scripts. It can be either way, of course my preference is to keep them to simplify it for others and to have people see the files that the scripts created. Also, if we need to run the scripts again incase of a bug or other, it would require less work on our part to run everything again

GD-Dheer replied:
We dont need to keep these since we are only running them once to populate the DB. Any technically inclined user will be able to run the scripts in sequential order here

JeremiahTee replied:
I would keep at least **datasets-with-datapoints.json** as it's been used many times to check against errors in the database or to verify that all the data has been parsed / inserted correctly. 
We also use websites like [JSON Viewer](http://jsonviewer.stack.hu/) to view this JSON is an organized way.
Since this JSON contains the 387 datasets with their corresponding data points, it might be used in the future for debugging purposes. In short we can keep that .json for accessibility and future testing purposes

GD-Dheer commented on Database/Scripts/Builders/authors-publications-builder.py:
```diff
@@ -0,0 +1,87 @@
+        tempCobject["name"] = x.replace(".\n", "")
+
+#dumping the dictionaries into the appropriate files.
+with open('./Scraped JSONs/authors.json', 'w') as outfile:
```
Im not sure if this is a windows thing but ./ will crash other os systems.

For portability:
import os
os.chdir('..')

TiraFire replied:
I will look into the portability, sounds like a good idea to make this work for everyone

GD-Dheer commented on Database/Scripts/Builders/datapoints-builder.py:
```diff
@@ -0,0 +1,97 @@
+          units = vari[1].replace("(", "").replace(")", "")
+          if "bar" in units or "torr" in units or "atm" in units :
+            variableNames.append("skip")
+            print(units)
```
Do we need to print here?

TiraFire replied:
Nope, was testing purposes, will remove

GD-Dheer commented on Database/Scripts/Builders/datapoints-builder.py:
```diff
@@ -0,0 +1,97 @@
+    data[key]["dataset-comment"] = ""
+    datasetIdCount += 1
+    #This opens the txt file with the datapoints per dataset
+    datasetfile = open("Builders/txt with datapoints/" + key + ".txt", "r")
```
Path problem. Will crash here.
Either use OS module or ../Builders/txt...

TiraFire replied:
Sounds good, will do

GD-Dheer commented on Database/Scripts/Queries in txt/1 - authors-query.txt:
```diff
@@ -0,0 +1,189 @@
+ INSERT INTO `databoom_db`.`authors` (`id`, `firstName`, `lastName`, `middleName`) VALUES (1, "S","Abid", "");
```
Similarly here, it seems like your /Scripts for SQL queries/ will generate all these txt files. We probably only need to keep default-queries

TiraFire replied:
Overall, we can remove them, it is a personal preference to keep them, gives a better idea to those looking at our repo what the scripts give as well as to use them without the need to run the scripts themselves (for reviewal process, these texts could be ignored)

Overall, the only files in need of reviewal would be the scripts as well as the builders IMO

GD-Dheer replied:
Its up to you. We have documentation that describe how the scripts work. The general idea is that we should commit code to the repo that is mandatory, these seem to be generated easily enough.

JeremiahTee replied:
Agreed, thank you for pointing that out. We can remove the 'text file queries'

GD-Dheer reviewed (changes requested):
Theres tons of great and hard work here. Nice job creating these scripts, well documented and coded.

I think we can cut down the committed files here pretty heavily. Also the paths need to be updated.. I had to manually modify them to run the scripts

TiraFire committed:
[#148] Fixed some duplicates in Authors as well as a bug in the script

Sean noticed a duplicate entry, turns out it was a bug in the script. Fixed it as well as a couple other authors i noticed duplicated.

TiraFire committed:
Testing Error with Pull

TiraFire committed:
Merge remote-tracking branch 'origin/Database/Populate-DB' into Database/Populate-DB

TiraFire committed:
[#148] Removed the queries, removed test print(), imported OS for files

removed the queries generated by the scripts (it explains how to generate them in the documentation), removed a small forgotten print message, and import OS library to help with portability when opening and creating files.

TiraFire committed:
[#148] Attempt at Fix for problem with OS.chdir

chdir worked with the cwd, instead of the path for the files, had to change it to work with the location of the file instead

JeremiahTee committed:
remove testing print statements

Removing these print statements as asked by GD. These were for testing purposes on Mac

GD-Dheer reviewed (approved):
Works perfectly! Great work
